### DISCUSSION TITLE
update wingetui itself from github

### DISCUSSION THREAD

**Discussion by SynapseScribe** [Devolutions/UniGetUI#1925] [General]
currently wingetui uses www.marticliment.com (if i am not wrong) to check for updates. why not update from github instead?

**Comment by marticliment** [comment]
Because using my own server allows me to customize how updates work, how they are checked, and to add a different endpoint.

Let's say, for example, my GitHub account gets hacked: I could still set the latest version to zero with an invalid hash and no one could publish an update with malware through the built-in updater.
If, let's say, my web server was hacked, the installer is still downloaded from GitHub so there would be no change to distribute an infected version.

**Comment by SynapseScribe** [comment]
so currently the update is done by checking on your website a link towards to github with a hash check, and the actual exe is actually downloaded from github? is this mechanism documented anywhere by curiosity?

**Reply by marticliment** [reply]
The link is a predefined link (it is hardcoded on WingetUI's source code) that points to the latest GitHub release, and the file used to retrieve the current version is this one: 

https://www.marticliment.com/versions/wingetui.ver

I am afraid there is no documentation about this mechanism, but if you are interested this is the function in charge of handling updates:
https://github.com/marticliment/WingetUI/blob/main/src/wingetui/Core/Tools.cs#L317C27-L317C51

**Comment by SynapseScribe** [comment]
Thanks for the clarifications, i think i better understand now.


I usually use and like open source apps for their simplicity and cleanness and because i am able to easilly identify endpoints (using some interactive firewall). Normally i am expecting the app to connect to the bare minimum endpoints needed to fulfil their jobs (in this case connecting to each of the package manager endpoints).


Some thoughts on the current approach:
- while the idea to delegate the updates check on a private website sounds more secure at first sight to check the sha256 hash, it might actually open other potentially more serious security and privacy concerns
	-- First of all, userbase ips are leaked to you (privacy concern) and presumably to a hacker in case your website gets hacked.
	-- i did an experiment blocking your website domain in system hosts, and observed that announcements don't work anymore. Supposedly, those are hosted on your website as well. In case your website would be hacked, maybe people could receive spam announcements
	-- handling the distribution and validation of update packages via a separate personal website is prone to potential misconfigurations or outdated or even vulnerable patches (if the website holds wrong hash / version)
	-- usually personal websites can be exposed to ddos.. in this case i see the website is under cloudflare which offers protection against it, so maybe it's not worthy mentioning
       -- a connection towards github is more trustable then to an unrecognized domain
	-- if hacking a 2fa protected account on github would be so easy, we'd have a much bigger problem worldwide
		-- From my observations so far, most of the self-updateable open source apps connect to github to do so
		-- it's far more hard to compromise a 2fa protected github account then to hack a personal website
			--- maybe it's wiser to trust microsoft/github in handling security hardening than do it yourself in a personal website. By transfering this responsability to them, the website as a dependency will not be needed anymore, reducing the needed maintenance and the area of attack
			
- BONUS: another idea other than connecting to github to self-update, would be to self-update via winget/chocolatey themselves and handle this particular case differently (launch installer, close wingetui itself, update, then reopen it automatically).. normally should be possible


I took a look in the code and even though I don't code in C# i was able to understand and was expecting this mechanism in the app. The missing puzzle is - How is the hash and version updated in the website page - is it done automatically or do you type it manually?
	- if done automatically by calculating the hash from the binary from github - it breaks the whole idea of crosschecking hash in case your github account would be compromised as in your concern (i see you're using github actions to build it, but didn't check the flows)
	- if you build it locally and automatically update the website page somehow, your pc could be prone to infection and binary compromised, and also its hash
	- if you enter it manually, it is prone to typos and time consuming
        - At first sight, none of the above approaches sounds good in the end. Or maybe I am missing something.. Checking for, and updating from github sounds like a better option imho.

**Reply by marticliment** [reply]
> I took a look in the code and even though I don't code in C# i was able to understand and was expecting this mechanism in the app. The missing puzzle is - How is the hash and version updated in the website page - is it done automatically or do you type it manually?

I type it manually. It is not that difficult nor annoying, I just update the version file on my local environment and then deploy the changes to the web server.

How would you implement a GH-based updater? I can't think of how to check for updates...

> * -- i did an experiment blocking your website domain in system hosts, and observed that announcements don't work anymore. Supposedly, those are hosted on your website as well. In case your website would be hacked, maybe people could receive spam announcements

I can easily invalidate the DNS records on my domain until I can fix it. The fact that I have my own server gives me more flexibility on how to handle updates and announcements. And while users could be *spammed* with advertising, the banner does only show on the settings window and does not have any notification capability nor anything else.

**Comment by SynapseScribe** [comment]
I asked AI for an example of such self-update implementation. I think it should be usable as guidance.


_There are a few options available for implementing a self-update mechanism for a C# Windows desktop application using GitHub Releases. Here's an example of how you might implement such a mechanism using GitHub Releases and Squirrel.Windows:_

    First, create a new GitHub repository for your application and upload your compiled executable file as a release asset. Make note of the URL for the latest release asset download link.
    Next, add the Squirrel.Windows NuGet package to your project by running Install-Package squirrel.windows in the Package Manager Console.
    Modify your application's startup code to perform the following tasks:
        Check the current version of the installed application.
        Send a request to the GitHub API to retrieve information about the latest release. You can do this using the HttpClient class and parsing the JSON response to extract the assets array.
        Compare the current version to the latest version retrieved from GitHub.
        If a newer version is available, prompt the user to download and install the update. You can use the Squirrel.Windows UpdateManager class to initiate the update process.



Self-updating a C# Windows desktop app can be implemented safely by having the app call an external updater program, which handles the update process separately from the main app. This approach ensures that the update process does not interfere with the functioning of the main app. Additionally, some developers recommend maintaining accurate DLL versions and ensuring that the updater can update itself. Overall, while developing a custom self-updating mechanism requires careful consideration and testing, it appears to be a viable option for managing software updates. However, it is essential to keep security best practices in mind during development and implementation to avoid potential vulnerabilities.


Here's an example of what the updated start code might look like:

```
using System;
using System.Diagnostics;
using System.IO;
using System.Net.Http;
using Newtonsoft.Json.Linq;

namespace MyApp
{
    static class Program
    {
        const string repoOwner = "myusername";
        const string repoName = "myrepositoryname";
        const string apiUrlBase = $"https://api.github.com/repos/{repoOwner}/{repoName}/releases/latest";
        
        static void Main()
        {
            // Get the current version number
            var exePath = Process.GetCurrentProcess().MainModule.FileName;
            var verInfo = FileVersionInfo.GetVersionInfo(exePath);
            var currentVersion = new SemanticVersion(verInfo.FileVersion);
            
            try
            {
                // Retrieve the latest release info from GitHub
                var client = new HttpClient();
                var jsonResponse = client.GetStringAsync(apiUrlBase).Result;
                var jObj = JObject.Parse(jsonResponse);
                var tagName = jObj["tag_name"].Value<string>();
                var assetsArray = jObj["assets"];
                
                // Find the latest release asset download link
                var latestReleaseAsset = assetsArray[0];
                var url = latestReleaseAsset["browser_download_url"].Value<string>();
                var fileName = Path.GetFileName(url);
                
                // Download the latest release asset
                var tempFilePath = Path.Combine(Path.GetTempPath(), fileName);
                using (var wc = new WebClient())
                {
                    wc.DownloadFile(url, tempFilePath);
                }
                
                // Extract the version number from the downloaded file name
                var semVerRegex = new System.Text.RegularExpressions.Regex(@"\d+\.\d+\.\d+");
                var match = semVerRegex.Match(fileName);
                var latestVersionStr = match.Value;
                var latestVersion = new SemanticVersion(latestVersionStr);
                
                // Prompt the user to update if necessary
                if (latestVersion > currentVersion)
                {
                    DialogResult result = MessageBox.Show($"A new version ({latestVersion}) of the application is available. Do you want to update now?", "New Version Available", MessageBoxButtons.YesNo);
                    if (result == DialogResult.Yes)
                    {
                        // Use Squirrel.Windows to install the update
                        var updateManager = new UpdateManager(tempFilePath);
                        updateManager.SetInstallFolder(AppDomain.CurrentDomain.BaseDirectory);
                        updateManager.BeginUpdate();
                        return;
                    }
                }
            }
            catch (Exception ex)
            {
                // Handle exceptions as needed
            }

            // Continue with normal application startup...
        }
    }
}
```



...But what about my other idea, to self update from within wingetUI itself via winget/chocolatey?

**Comment by SynapseScribe** [comment]
Regarding hash check, AI replied:

Assuming that you are using GitHub releases to distribute your application updates, one way to ensure app integrity through hash checks is to include a precomputed hash value alongside each released executable. This allows users to validate the authenticity and integrity of the downloaded file by checking its corresponding hash value after the download completes.

To achieve this, you could follow these general steps:

    Choose a suitable hashing algorithm, such as SHA-256 or SHA-512, that offers sufficient security and resistance to collisions.
    Compute the hash value for the final build of your application binary before packaging and releasing it on GitHub.
    Upload both the packaged binary and its associated hash value along with the release on GitHub. Ideally, you could provide them together in a single archive or attach them separately to the release page.
    Upon receiving the update notification, download the appropriate binary and hash value from the latest GitHub release.
    Calculate the hash value locally for the downloaded binary using the same chosen hashing algorithm.
    Compare the calculated local hash value with the received hash value. If they match, proceed with the installation. Otherwise, abort the update process and notify the user about a potential compromise or corruption.

By incorporating hash checks into your self-update mechanism, you can enhance the overall security posture of your application distribution pipeline and improve end-user trust. Keep in mind that including regular security audits and practicing good coding hygiene throughout your development cycle will further strengthen the robustness of your solution.

**Comment by marticliment** [comment]
Will investigate this, but I am not sure if I will change the updater
